daemon/cdi_linux.go

```diff
@@ -0,0 +1,42 @@
+package daemon
+
+import (
+	"fmt"
+
+	"github.com/docker/docker/pkg/capabilities"
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+)
+
+func registerCdiDriver() error {
+	driver := &deviceDriver{
+		capset:     capabilities.Set{"cdi": struct{}{}},
+		updateSpec: injectCdiDevices,
+	}
+
+	registerDeviceDriver("cdi", driver)
+
+	return nil
+}
+
+// injectCdiDevices is a updateSpecHander for the CDI device driver
+func injectCdiDevices(s *specs.Spec, dev *deviceInstance) error {
+	return injectCDI(s, dev.req.DeviceIDs)
+}
+
+// injectCDI injects a CDI device into the OCI spec.
+func injectCDI(s *specs.Spec, cdiDeviceNames []string) error {
+	if len(cdiDeviceNames) == 0 {
+		return nil
+	}
+
+	if cdiRegistry == nil {
+		return fmt.Errorf("CDI registry not initialized")
+	}
+
+	_, err := cdiRegistry.InjectDevices(s, cdiDeviceNames...)
```

Comment:
Does this auto-refresh the registry, i.e. to handle CDI Specs added (or removed) after startup?

Reply:
Yes, the default mode for the registry is to watch the configured folders (`/etc/cdi` and `/var/run/cdi` by default) and update the cached specs on changes. This means that the calls to `InjectDevices` should always have access to up-to-date specs.